plugins/sh/src/com/intellij/sh/lexer/Sh.flex

```diff
@@ -248,14 +251,32 @@ EvalContent              = [^\r\n$\"`'() ;] | {EscapedChar}
 <CONDITIONAL_EXPRESSION> {
     "=="                          { return EQ; }
     "!="                          { return NE; }
-    "=~"                          { pushState(REGULAR_EXPRESSION); return REGEXP; }
+    "=~"                          { regexStart = getTokenEnd(); regexGroups = 0; pushState(REGULAR_EXPRESSION); return REGEXP; }
     "<"                           { return LT; }
     ">"                           { return GT; }
 }
 
 <REGULAR_EXPRESSION> {
-  {WhiteSpace}+                   { return WHITESPACE;}
-  {RegexInQuotes}                 { popState(); return WORD; }
+  "$(("                              { isArithmeticExpansion = true; yypushback(2); return DOLLAR; }
```

Comment:
`REGULAR_EXPRESSION` duplicates `PARAMETER_EXPANSION`. I prefer to combine them

Reply:
There are duplicate rules for `$((`, `$(`, `${`, `$[`, `{Variable}`, `{Quote}`, `{RawString}`, `{UnclosedRawString}. The other rules are different.

I could extract the shared rules into a seperate block above. Is that what you were suggesting?

Reply:
Yes, I think we can do something like this:
```
< REGULAR_EXPRESSION > {
...
}
<PARAMETER_EXPANSION > {
...
}
< REGULAR_EXPRESSION , PARAMETER_EXPANSION > {
"$(("                              { isArithmeticExpansion = true; yypushback(2); return DOLLAR; }
"$("                               { pushState(PARENTHESES_COMMAND_SUBSTITUTION); yypushback(1); return DOLLAR; }
"${"                               { pushState(PARAMETER_EXPANSION); yypushback(1); return DOLLAR;}
...
}
```

Reply:
Done. I had to move the shared rules before the other groups to keep the previous order how the rules are applied.